lib/open_api_spex/open_api/map.ex

```diff
@@ -0,0 +1,395 @@
+defmodule OpenApiSpex.OpenApi.Map do
+  @moduledoc """
+  This module exposes functionality to convert an arbitrary map into a OpenApi struct.
+  """
+  alias OpenApiSpex.{
+    Components,
+    Contact,
+    Discriminator,
+    Encoding,
+    Example,
+    ExternalDocumentation,
+    Header,
+    Info,
+    License,
+    Link,
+    MediaType,
+    OAuthFlow,
+    OAuthFlows,
+    OpenApi,
+    Operation,
+    Parameter,
+    PathItem,
+    Reference,
+    RequestBody,
+    Response,
+    Schema,
+    SecurityScheme,
+    Server,
+    ServerVariable,
+    Tag,
+    Xml
+  }
+
+  def decode(%{openapi: _openapi, info: _info, paths: _paths} = map) do
+    map
+    |> to_struct(OpenApi)
+    |> to_struct(:info, Info)
+    |> to_struct(:paths, PathItems)
+    |> to_struct(:servers, Servers)
+    |> to_struct(:components, Components)
+    |> to_struct(:tags, Tag)
+    |> to_struct(:externalDocs, ExternalDocumentation)
+  end
+
+  defp embedded_ref_or_struct(list, mod) when is_list(list) do
+    list
+    |> Enum.map(fn
+      %{"$ref": _} = v -> struct(Reference, v)
+      v -> to_struct(v, mod)
+    end)
+  end
+
+  defp embedded_ref_or_struct(map, mod) when is_map(map) do
+    map
+    |> Map.new(fn
+      {k, %{"$ref": _} = v} ->
+        {k, struct(Reference, v)}
+
+      {k, v} ->
+        {k, to_struct(v, mod)}
+    end)
+  end
+
+  defp convert_value_to_atom_if_present(map, key) do
+    if Map.has_key?(map, key) do
+      Map.update!(map, key, fn
+        nil ->
+          nil
+
+        v ->
+          String.to_atom(v)
+      end)
+    else
+      map
+    end
```

Comment:
This would be a good function to use pattern matching:

```elixir
case map do
  %{^key => nil} -> nil
  %{^key => value} -> String.to_atom(value)
  %{} -> map
end
```

But as I said above, there is something awkward about the return type of this function. :)